### PR TITLE
added embedded video

### PR DESCRIPTION
**Pull request by weining-li**
Added support for embedded Youtube and Vimeo videos
#43

### CODE REVIEW

**Review by clpetersonucf — CHANGES_REQUESTED** [review]
Played around with it, it works well and appears totally functional so far, a few feedback items:

- The new media upload dialog could use some additional styling and information about the choice presented to the user - I can help with that if needed. It's very barebones currently.
- The new upload dialog doesn't have a cancel option, you either have to enter a [valid] video URL, or hit the upload image button and cancel from there
- The video upload input doesn't appear to handle invalid urls or strings very well (I got a console error)
- Similarly, the last item entered in the video input URL persists when the dialog is closed and then opened again for a different node, even if the string was invalid. It should probably be cleared
- Selecting "upload image" appears to pre-populate the media target with whatever the last media item chosen was - either a video, or the last selected image. Cancelling out from the regular Materia media uploader appears to cause an unintended state, like this example with a MC node:
Screen Shot 2020-10-27 at 4 38 22 PM

- Hotspot nodes bring up the new intermediate dialog by default. Hotspot nodes are a special case, we should limit these to just images.
Screen Shot 2020-10-27 at 4 40 02 PM

- This issue may not have originated with your PR, but there appears to be scaling issues with media that's located to the left of question text. With youtube videos it's especially evident. We should look at making it a little more consistent. Example:
 
Screen Shot 2020-10-27 at 4 44 16 PM

**Comment by clpetersonucf** [comment]
Everything looks good so far. Styling the media selection window will really help to put finishing touches on this. The only thing I'm noticing:

- The drag handles for elliptical and square hotspot labels don't appear to be working right now. I can't interact with them at all, like their Z-index may have changed, or perhaps there's an invisible element above that that's blocking them. To confirm, I tried it on master and the problem didn't appear evident.

Media that's on the left side still appears to occupy too much of the horizontal width of the content area, but that's an issue that appears on master, so we can let it ride for the purposes of this pull request.

File: src/creator.html [hunk 1]
```diff
@@ -676,7 +678,8 @@ <h4>Manage This Hotspot</h4>
 						<!-- Background image placed behind the hotspot's SVG canvas -->
 						<img id="hotspotImage"
 							ng-src="{{image.src}}"
-							ng-show="mediaReady" />
+							ng-show="mediaReady && showImage" />
+						<iframe ng-show="mediaReady && !showImage" width="100%" height="100%" id="hotspotImage" ng-src="{{url}}" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```
Comment (clpetersonucf) [inline src/creator.html:682]: Is this iframe still required for hotspots?

**Review by clpetersonucf — CHANGES_REQUESTED** [review]

File: src/creator.html (hunk 1 above)
Comment (weining-li) [reply src/creator.html:682]: oops. It shouldn't be, I'll remove it

Comment (clpetersonucf) [reply src/creator.html]: :+1:

**Comment by clpetersonucf** [comment]
Tested on QA and found an additional issue:

Asset information is being saved in the qset as a media type, e.g.,
```
"asset": {
                        "materiaType": "asset",
                        "align": "right",
                        "id": "https://www.youtube.com/embed/qJRKedSUHg4",
                        "type": "video"
                    }
```
but by default materia prepends the media url to the asset id when creating the actual asset url. That causes the resulting url for the embedded video to look something like this:

```
https://materia.qa.cdl.ucf.edu/media/https://www.youtube.com/embed/97GrgHKg47c
```

IMO, this probably shouldn't be considered a true materia asset at all, since there isn't a corresponding asset being saved to the database, and having the URL act as the media ID is ugly and could cause database collisions. If anything, the asset id should be provided by the DB, like other asset types, and there should be another property specifically for the video url.

**Comment by clpetersonucf** [comment]
Still breaking on QA when I try to embed a video, save the qset, then reload and revisit the node editor where the video was embedded.

I think what's happening: there's one place where you overlooked how the widget is handling media, when the qset is loaded in the creator and converted into the tree. There's currently nothing to differentiate how images and video assets are handled, so it's trying to generate a media URL for video assets and failing to do so. Check `createTreeDataFromQset` in `creator-assets/services.coffee`

**Review by clpetersonucf — CHANGES_REQUESTED** [review]
We're nearly there. Everything looks good and saving/loading/playing qsets work fine, I'm just noticing some verbose console log statements that were left in. Once those are removed this looks good to go. :+1:

**Comment by weining-li** [comment]
oops, removed.

**Review by clpetersonucf — APPROVED** [review]

**Pull request merged**